### ISSUE TITLE
TTL callback is invoked with significant delay 

### ISSUE THREAD

**Issue by alexterman** [hazelcast/hazelcast#8894]
Dear Hazelcast folks,

I'm using Hazelcast 3.7.1 and trying to use TTL + eviction listener.
In general eviction listener called with significant delay.
I configured setProperty("hazelcast.internal.map.expiration.task.period.seconds", "1"); as it was suggested in the #7722 issue.

I inserted 100 elements to the map and set TTL to 5. Here the [gist with logs](https://gist.github.com/alexterman/cc39288cbb8e01cda3bcdfd618de114d)
As I can see from the prints my eviction callback called in batches of around 15 items 7 -8 times and the max delay is stands for 12 seconds. 

I was expecting 1 batch of all items.
Should I configure more stuff to get it work?

Thanks,
Alex

**Comment by ahmetmircik** [comment]
Hi @alexterman,

These are the most aggressive settings. You retry by playing them.

```
System.setProperty("hazelcast.internal.map.expiration.task.period.seconds", "1");
System.setProperty("hazelcast.internal.map.expiration.cleanup.percentage", "100");<-- makes full table scan in a partition to find expirable entries.
System.setProperty("hazelcast.internal.map.expiration.cleanup.operation.count", "271"); <-- sets cleanup operation count to partition count. 271 is current default. Otherwise it will pick a small value which doesn't hurt the system.
```

**Comment by alexterman** [comment]
Hi @ahmetmircik ,

Thanks for your reply.
After adding the suggested parameters situation improved right a way.
Seems like all entries evicted in one batch and max delay stands for 2 seconds.
A few things I noticed:
1. After increasing number of inserted entries to 1 million ~20% of entries were delayed, but still 2 seconds delay (delay is not increasing).
2. The parameters , you mentioned, works only if configuring them in a System level and not instance level (which means all instances will be forced to use those settings).
3. I couldn't find documentation describing those parameters in relation to eviction and TTL.

Thanks,
Alex

**Comment by ahmetmircik** [comment]
we will try to address 2 and 3 in 3.8

**Comment by alexterman** [comment]
Great, thanks.

**Comment by ahmetmircik** [comment]
we will close it after improvements done, now keeping it open

**Comment by ahmetmircik** [comment]
2 is fixed via https://github.com/hazelcast/hazelcast/pull/8907
3 is fixed via https://github.com/hazelcast/hazelcast/pull/8950

These fixes will be available in 3.8.
